Python/optimizer_analysis.c

```diff
@@ -1,20 +1,211 @@
+     * know that the globals dict is the one we expected, and
+     * that it hasn't changed
+     * In order to treat builtins as constants,  we need to
+     * know that the builtins dict is the one we expected, and
+     * that it hasn't changed and that the global dictionary's
+     * keys have not changed */
+    uint32_t builtins_checked = 0;
+    uint32_t builtins_watched = 0;
+    uint32_t globals_checked = 0;
+    uint32_t globals_watched = 0;
+    if (interp->dict_state.watchers[1] == NULL) {
+        interp->dict_state.watchers[1] = globals_watcher_callback;
+    }
+    for (int pc = 0; pc < buffer_size; pc++) {
+        _PyUOpInstruction *inst = &buffer[pc];
+        int opcode = inst->opcode;
+        switch(opcode) {
+            case _GUARD_BUILTINS_VERSION:
+                if (incorrect_keys(inst, builtins)) {
+                    return 0;
+                }
+                if (interp->rare_events.builtin_dict >= _Py_MAX_ALLOWED_BUILTINS_MODIFICATIONS) {
+                    continue;
+                }
+                if ((builtins_watched & 1) == 0) {
+                    PyDict_Watch(0, builtins);
+                    builtins_watched |= 1;
+                }
+                if (builtins_checked & 1) {
+                    buffer[pc].opcode = NOP;
+                }
+                else {
+                    buffer[pc].opcode = _CHECK_BUILTINS;
+                    buffer[pc].operand = (uintptr_t)builtins;
+                    builtins_checked |= 1;
+                }
+                break;
+            case _GUARD_GLOBALS_VERSION:
+                if (incorrect_keys(inst, globals)) {
+                    return 0;
+                }
+                uint64_t watched_mutations = get_mutations(globals);
+                if (watched_mutations >= _Py_MAX_ALLOWED_GLOBALS_MODIFICATIONS) {
+                    continue;
+                }
+                if ((globals_watched & 1) == 0) {
+                    PyDict_Watch(1, globals);
+                    _Py_BloomFilter_Add(dependencies, globals);
+                    globals_watched |= 1;
+                }
+                if (globals_checked & 1) {
+                    buffer[pc].opcode = NOP;
+                }
+                else {
+                    buffer[pc].opcode = _CHECK_GLOBALS;
+                    buffer[pc].operand = (uintptr_t)globals;
+                    globals_checked |= 1;
+                }
+                break;
+            case _LOAD_GLOBAL_BUILTINS:
+                if (globals_checked & builtins_checked & globals_watched & builtins_watched & 1) {
+                    global_to_const(inst, builtins);
+                }
+                break;
+            case _LOAD_GLOBAL_MODULE:
+                if (globals_checked & globals_watched & 1) {
+                    global_to_const(inst, globals);
+                }
+                break;
+            case _JUMP_TO_TOP:
+            case _EXIT_TRACE:
+                return 1;
```

Comment:
Optimizers might add more instructions, and possibly change the order, so I think we want to check for exits, rather than count uops.
I've moved them to the end.

We should only need one jump back and one unconditional exit, so this shouldn't get more complicated.